Source/DafnyTests/DafnyTestCase.cs

```diff
@@ -0,0 +1,417 @@
+  // can we move all the dafny-specific stuff to DafnyTestSpec instead?
+  public class DafnyTestCase: IXunitSerializable {
+
+    public string[] Arguments;
+
+    public class Expectation : IXunitSerializable {
+      // May be null if the test doesn't need to check the output
+      public string OutputFile;
+      public int ExitCode = 0;
+
+      public string SpecialCaseReason;
+
+      public static Expectation NO_OUTPUT = new Expectation(0, null, null);
+      
+      public Expectation(int exitCode, string outputFile, string specialCaseReason) {
+        ExitCode = exitCode;
+        OutputFile = outputFile;
+        SpecialCaseReason = specialCaseReason;
+      }
+
+      public Expectation() {
+        
+      }
+      
+      public void Deserialize(IXunitSerializationInfo info) {
+        OutputFile = info.GetValue<string>(nameof(OutputFile));
+        ExitCode = info.GetValue<int>(nameof(ExitCode));
+        SpecialCaseReason = info.GetValue<string>(nameof(SpecialCaseReason));
+      }
+
+      public void Serialize(IXunitSerializationInfo info) {
+        info.AddValue(nameof(OutputFile), OutputFile);
+        info.AddValue(nameof(ExitCode), ExitCode);
+        info.AddValue(nameof(SpecialCaseReason), SpecialCaseReason);
+      }
+
+      public Expectation Adjust() {
+        return new Expectation(ExitCode, OutputFile == null ? null : "comp/" + OutputFile, SpecialCaseReason);
+      }
+      
+      public override string ToString() {
+        return OutputFile ?? "-";
+      }
+    }
+
+    public Expectation Expected;
+
+    public DafnyTestCase(IEnumerable<string> arguments, Expectation expected) {
+      Expected = expected;
+      Arguments = arguments.ToArray();
+    }
+
+    public DafnyTestCase() {
+      
+    }
+  
+    public void Serialize(IXunitSerializationInfo info) {
+      info.AddValue(nameof(Arguments), Arguments);
+      info.AddValue(nameof(Expected), Expected);
+    }
+    
+    public void Deserialize(IXunitSerializationInfo info) {
+      Arguments = info.GetValue<string[]>(nameof(Arguments));
+      Expected = info.GetValue<Expectation>(nameof(Expected));
+    }
+
+    public static ProcessResult RunDafny(IEnumerable<string> arguments) {
+      // TODO-RS: Let these be overridden
+      List<string> dafnyArguments = new List<string> {
+        // Expected output does not contain logo
+        "/nologo",
+        "/countVerificationErrors:0",
+
+        // We do not want absolute or relative paths in error messages, just the basename of the file
+        "/useBaseNameForFileName",
+
+        // We do not want output such as "Compiled program written to Foo.cs"
+        // from the compilers, since that changes with the target language
+        "/compileVerbose:0"
+      };
+      dafnyArguments.AddRange(arguments);
+
+      using (Process dafnyProcess = new Process()) {
+        dafnyProcess.StartInfo.FileName = DafnyTestSpec.DAFNY_EXE;
+        foreach(var argument in dafnyArguments) {
```

Comment:
np to match other uses of `foreach` in this file (also just generally)
```suggestion
        foreach (var argument in dafnyArguments) {
 ```